common/tools/dev-tool/src/util/migrations.ts

```diff
@@ -0,0 +1,609 @@
+ */
+function findFirstPendingMigrationIndex(migrations: readonly Migration[], date: Date): number {
+  let low = 0;
+  let high = migrations.length - 1;
+
+  while (low <= high) {
+    const mid = Math.floor((low + high) / 2);
+    const m = migrations[mid];
+
+    if (isPending(m, date)) {
+      if (mid === 0 || !isPending(migrations[mid - 1], date)) return mid;
+      high = mid - 1;
+    } else {
+      low = mid + 1;
+    }
+  }
+
+  return migrations.length;
+}
+
+/**
+ * Get an iterable of all migrations that are pending relative to a `date`.
+ *
+ * @param date - a `migrationDate` to compare against
+ */
+export function* listPendingMigrations(date: Date): Iterable<Migration> {
+  const migrations = getSortedMigrations();
+  const firstIndex = findFirstPendingMigrationIndex(migrations, date);
+
+  for (const m of migrations.slice(firstIndex)) {
+    if (isPending(m, date)) yield m;
+  }
+}
+
+/**
+ * @param migration - the migration in question
+ * @param date - a `migrationDate` to compare to
+ * @returns true if the migration is considered 'applied'
+ */
+export function isApplied(migration: Migration, date: Date): boolean {
+  return !isPending(migration, date);
+}
+
+/**
+ * Get an iterable of all migrations that have been applied relative to a `date`.
+ *
+ * @param date - a `migrationDate` to compare against
+ */
+export function* listAppliedMigrations(date: Date): Iterable<Migration> {
+  const migrations = getSortedMigrations();
+
+  for (const migration of migrations) {
+    if (isApplied(migration, date)) {
+      yield migration;
+    } else break;
+  }
+}
+
+/**
+ * Gets a migration by its ID. This requires that the migration be registered using `createMigration`.
+ *
+ * @param id - the ID of the migration to search for.
+ * @returns
+ */
+export function getMigrationById(id: string): Migration | undefined {
+  return MIGRATIONS_BY_ID.get(id);
+}
+
+let LOADED = false;
+let _loadJob: Promise<void>;
+export function loadMigrations() {
+  if (LOADED) return Promise.resolve();
+
+  return (_loadJob ??= (async function () {
+    for await (const f of findMatchingFiles(
+      path.resolve(__dirname, "..", "migrations"),
+      (fn) => fn.endsWith(".ts") && !fn.endsWith(".d.ts")
+    )) {
+      await import(f);
+    }
+  })().then(() => {
+    LOADED = true;
+  }));
+}
+
+// #region suspensions
+
+export interface SuspendedMigrationState {
+  id: string;
+  path: string;
+  phase: "execute" | "validate";
+}
+
+const STATE_PATH_SUFFIX = ["azsdk-dev-tool", "state", "migration-suspended"];
+
+let _stateFile: string;
+
+/**
+ * Gets the path to the state file. This function is cached and the result will only be computed once.
+ *
+ * @returns the OS-appropriate file path where a migration state file may be stored
+ */
+function getStateFilePath(): string {
+  return (_stateFile ??= _getStateFilePath());
+}
+
+/**
+ * Helper for computing the stateFile path.
+ * @internal
+ */
+function _getStateFilePath(): string {
```

Comment:
Yeah.. this is a good question. I was going to re-use env-paths since it was already being used for test-proxy but the only packages I found for this were very underutilized so I figured I'd just implement it. We could centralize the env-paths + this logic into a util module for host paths?

Reply:
Centralizing it sounds like a good idea. If we're hand-rolling the state path calculation, I would lean toward cutting out `env-paths` entirely and using similar logic for the cache directory too.

Reply:
Added https://github.com/Azure/azure-sdk-for-js/issues/25728 to track.